File: src/main/java/org/bsplines/ltexls/languagetool/LanguageToolJavaInterface.java
package org.bsplines.ltexls.languagetool;

import java.io.File;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.concurrent.TimeUnit;

import javax.xml.parsers.ParserConfigurationException;

import org.bsplines.ltexls.Tools;

import org.checkerframework.checker.nullness.qual.EnsuresNonNullIf;
import org.checkerframework.checker.nullness.qual.MonotonicNonNull;

import org.languagetool.AnalyzedSentence;
import org.languagetool.AnalyzedTokenReadings;
import org.languagetool.JLanguageTool;
import org.languagetool.Language;
import org.languagetool.Languages;
import org.languagetool.ResultCache;
import org.languagetool.UserConfig;
import org.languagetool.markup.AnnotatedText;
import org.languagetool.rules.Rule;
import org.languagetool.rules.RuleMatch;
import org.languagetool.rules.patterns.AbstractPatternRule;

import org.xml.sax.SAXException;

public class LanguageToolJavaInterface extends LanguageToolInterface {
  private @MonotonicNonNull JLanguageTool languageTool;

  private static final int resultCacheExpireAfterMinutes = 10;

  /**
   * Constructor.
   *
   * @param languageShortCode short code of the checking language
   * @param motherTongueShortCode short code of the mother tongue language
   * @param sentenceCacheSize size of the sentence cache in sentences
   * @param dictionary list of words of the user dictionary
   */
  public LanguageToolJavaInterface(String languageShortCode, String motherTongueShortCode,
        int sentenceCacheSize, List<String> dictionary) {
    if (!Languages.isLanguageSupported(languageShortCode)) {
      Tools.logger.severe(Tools.i18n("notARecognizedLanguage", languageShortCode));
      return;
    }

    Language language = Languages.getLanguageForShortCode(languageShortCode);
    Language motherTongue = ((!motherTongueShortCode.isEmpty())
        ? Languages.getLanguageForShortCode(motherTongueShortCode) : null);
    ResultCache resultCache = new ResultCache(sentenceCacheSize,
        resultCacheExpireAfterMinutes, TimeUnit.MINUTES);
    UserConfig userConfig = new UserConfig(dictionary);

    @SuppressWarnings("argument.type.incompatible")
    JLanguageTool languageTool = new JLanguageTool(language, motherTongue, resultCache, userConfig);
    this.languageTool = languageTool;
  }

  @EnsuresNonNullIf(expression = "this.languageTool", result = true)
  @Override
  public boolean isReady() {
    return (languageTool != null);
  }

  @Override
  public List<LanguageToolRuleMatch> check(AnnotatedText annotatedText) {
    if (!isReady()) {
      Tools.logger.warning(Tools.i18n("skippingTextCheck"));
      return Collections.emptyList();
    }

    List<RuleMatch> matches;

    try {
      matches = languageTool.check(annotatedText);
    } catch (RuntimeException | IOException e) {
      Tools.logger.severe(Tools.i18n("languageToolFailed", e.getMessage()));
      e.printStackTrace();
      return Collections.emptyList();
    }

    List<LanguageToolRuleMatch> result = new ArrayList<>();

    for (RuleMatch match : matches) {
      int fromPos = match.getFromPos();
      int toPos = match.getToPos();
      String ruleId = ((match.getRule() != null) ? match.getRule().getId() : null);
      String message = match.getMessage();
      String sentence = ((match.getSentence() != null) ? match.getSentence().getText() : null);
      List<String> suggestedReplacements = match.getSuggestedReplacements();
      result.add(new LanguageToolRuleMatch(ruleId, sentence, fromPos, toPos, message,
          suggestedReplacements));
    }

    return result;
  }

  @Override
  public void activateDefaultFalseFriendRules() {
    if (!isReady()) return;

    // from JLanguageTool.activateDefaultFalseFriendRules (which is private)
    String falseFriendRulePath = JLanguageTool.getDataBroker().getRulesDir() + "/"
        + JLanguageTool.FALSE_FRIEND_FILE;

    try {
      List<AbstractPatternRule> falseFriendRules = languageTool.loadFalseFriendRules(
          falseFriendRulePath);
      for (Rule rule : falseFriendRules) languageTool.addRule(rule);
    } catch (ParserConfigurationException | SAXException | IOException e) {
      Tools.logger.warning(Tools.i18n("couldNotLoadFalseFriendRules",
          falseFriendRulePath, e.getMessage()));
      e.printStackTrace();
    }
  }

  @Override
  public void activateLanguageModelRules(String languageModelRulesDirectory) {
    if (!isReady()) return;

    try {
      languageTool.activateLanguageModelRules(new File(languageModelRulesDirectory));
    } catch (IOException | RuntimeException e) {
      Tools.logger.warning(Tools.i18n("couldNotLoadLanguageModel",
          languageModelRulesDirectory, e.getMessage()));
      e.printStackTrace();
    }
  }

  @Override
  public void activateNeuralNetworkRules(String neuralNetworkRulesDirectory) {
    if (!isReady()) return;

    try {
      languageTool.activateNeuralNetworkRules(new File(neuralNetworkRulesDirectory));
    } catch (IOException | RuntimeException e) {
      Tools.logger.warning(Tools.i18n("couldNotLoadNeuralNetworkModel",
          neuralNetworkRulesDirectory, e.getMessage()));
      e.printStackTrace();
    }
  }

  @Override
  public void activateWord2VecModelRules(String word2vecRulesDirectory) {
    if (!isReady()) return;

    try {
      languageTool.activateWord2VecModelRules(new File(word2vecRulesDirectory));
    } catch (IOException | RuntimeException e) {
      Tools.logger.warning(Tools.i18n("couldNotLoadWord2VecModel",
          word2vecRulesDirectory, e.getMessage()));
      e.printStackTrace();
    }
  }

  @Override
  public void enableRules(List<String> ruleIds) {
    if (!isReady()) return;

    // for strange reasons there is no JLanguageTool.enableRules
    for (String ruleId : ruleIds) {
      languageTool.enableRule(ruleId);
    }
  }

  @Override
  public void disableRules(List<String> ruleIds) {
    if (!isReady()) return;
    languageTool.disableRules(ruleIds);
  }

  @Override
  public void enableEasterEgg() {
    if (!isReady()) return;

    languageTool.addRule(new Rule() {
      public String getId() {
        return "bspline";
      }

      public String getDescription() {
        return "Unknown basis function";
      }

      public RuleMatch[] match(AnalyzedSentence sentence) {
        List<RuleMatch> matches = new ArrayList<>();
        for (AnalyzedTokenReadings token : sentence.getTokens()) {
          if (token.getToken().equalsIgnoreCase("hat")) {
            matches.add(new RuleMatch(this, sentence, token.getStartPos(), token.getEndPos(),
                "Unknown basis function. Did you mean <suggestion>B-spline</suggestion>?"));
          }
        }
        return matches.toArray(new RuleMatch[]{});
      }
    });

    languageTool.addRule(new Rule() {
      public String getId() {
        return "ungendered";
      }

      public String getDescription() {
        return "Ungendered variant";
      }

      public RuleMatch[] match(AnalyzedSentence sentence) {
        List<RuleMatch> matches = new ArrayList<>();
        for (AnalyzedTokenReadings token : sentence.getTokens()) {
          String s = token.getToken();
          if ((s.length() >= 2)
                && (s.substring(s.length() - 2, s.length()).equalsIgnoreCase("er"))) {
            matches.add(new RuleMatch(this, sentence, token.getStartPos(), token.getEndPos(),
                "Ungendered variant detected. "
                + "Did you mean <suggestion>" + s + "*in</suggestion>?"));
          }
        }
        return matches.toArray(new RuleMatch[]{});
      }
    });
  }
}

File: src/main/kotlin/org/bsplines/ltexls/languagetool/LanguageToolJavaInterface.kt
/* Copyright (C) 2019-2023 Julian Valentin, LTeX Development Community
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

package org.bsplines.ltexls.languagetool

import org.apache.commons.text.StringEscapeUtils
import org.bsplines.ltexls.parsing.AnnotatedTextFragment
import org.bsplines.ltexls.tools.I18n
import org.bsplines.ltexls.tools.Logging
import org.bsplines.ltexls.tools.Tools
import org.languagetool.AnalyzedSentence
import org.languagetool.AnalyzedTokenReadings
import org.languagetool.JLanguageTool
import org.languagetool.Language
import org.languagetool.Languages
import org.languagetool.ResultCache
import org.languagetool.RuleMatchListener
import org.languagetool.UserConfig
import org.languagetool.rules.Rule
import org.languagetool.rules.RuleMatch
import org.languagetool.rules.patterns.AbstractPatternRule
import org.xml.sax.SAXException
import java.io.File
import java.io.IOException
import java.io.OutputStream
import java.io.PrintStream
import java.util.concurrent.TimeUnit
import java.util.logging.Level
import javax.xml.parsers.ParserConfigurationException

class LanguageToolJavaInterface(
  languageShortCode: String,
  motherTongueShortCode: String,
  sentenceCacheSize: Long,
  dictionary: Set<String>,
) : LanguageToolInterface() {
  private val resultCache =
      ResultCache(sentenceCacheSize, RESULT_CACHE_EXPIRE_AFTER_MINUTES, TimeUnit.MINUTES)
  private val languageTool: JLanguageTool?

  init {
    if (Languages.isLanguageSupported(languageShortCode)) {
      val language: Language = Languages.getLanguageForShortCode(languageShortCode)
      val motherTongue: Language? = if (motherTongueShortCode.isNotEmpty()) {
        Languages.getLanguageForShortCode(motherTongueShortCode)
      } else {
        null
      }
      val userConfig = UserConfig(dictionary.toList())

      this.languageTool = JLanguageTool(language, motherTongue, this.resultCache, userConfig)
    } else {
      Logging.LOGGER.severe(I18n.format("notARecognizedLanguage", languageShortCode))
      this.languageTool = null
    }
  }

  override fun isInitialized(): Boolean {
    return (this.languageTool != null)
  }

  @Suppress("INACCESSIBLE_TYPE")
  override fun checkInternal(
    annotatedTextFragment: AnnotatedTextFragment,
  ): List<LanguageToolRuleMatch> {
    val languageTool: JLanguageTool? = this.languageTool

    if (languageTool == null) {
      Logging.LOGGER.warning(I18n.format("skippingTextCheckAsLanguageToolHasNotBeenInitialized"))
      return emptyList()
    }

    logResultCache()

    val ruleMatchListener: RuleMatchListener? = null
    val ruleLevel: JLanguageTool.Level =
    if (annotatedTextFragment.codeFragment.settings.enablePickyRules) {
      JLanguageTool.Level.PICKY
    } else {
      JLanguageTool.Level.DEFAULT
    }

    annotatedTextFragment.document.raiseExceptionIfCanceled()
    languageTool.setCheckCancelledCallback(
      JLanguageTool.CheckCancelledCallback {
        annotatedTextFragment.document.raiseExceptionIfCanceled()
        false
      },
    )

    val matches: List<RuleMatch> = try {
      // workaround bugs like https://github.com/languagetool-org/languagetool/issues/3181,
      // in which LT prints to stdout instead of stderr (this messes up the LSP communication
      // and results in a deadlock) => temporarily discard output to stdout
      val stdout: PrintStream = System.out
      System.setOut(
        PrintStream(
          object : OutputStream() {
            override fun write(b: Int) {
            }
          },
          false,
          "utf-8",
        ),
      )

      try {
        languageTool.check(
          annotatedTextFragment.annotatedText,
          true,
          JLanguageTool.ParagraphHandling.NORMAL,
          ruleMatchListener,
          JLanguageTool.Mode.ALL,
          ruleLevel,
        )
      } finally {
        System.setOut(stdout)
      }
    } catch (e: IOException) {
      Tools.rethrowCancellationException(e)
      Logging.LOGGER.severe(I18n.format("languageToolFailed", e))
      return emptyList()
    }

    annotatedTextFragment.document.raiseExceptionIfCanceled()
    val result = ArrayList<LanguageToolRuleMatch>()

    for (match: RuleMatch in matches) {
      result.add(LanguageToolRuleMatch.fromLanguageTool(match, annotatedTextFragment))
    }

    return result
  }

  @Suppress("INACCESSIBLE_TYPE")
  private fun logResultCache() {
    if (Logging.LOGGER.isLoggable(Level.FINER)) {
      Logging.LOGGER.finer("matchesCache.size() = " + this.resultCache.matchesCache.size())
      Logging.LOGGER.finer(
        "remoteMatchesCache.size() = " + this.resultCache.remoteMatchesCache.size(),
      )
      Logging.LOGGER.finer("sentenceCache.size() = " + this.resultCache.sentenceCache.size())

      if (Logging.LOGGER.isLoggable(Level.FINEST)) {
        Logging.LOGGER.finest(
          "matchesCache = " + mapToString(this.resultCache.matchesCache.asMap()),
        )
        Logging.LOGGER.finest(
          "remoteMatchesCache = " + mapToString(this.resultCache.remoteMatchesCache.asMap()),
        )
        Logging.LOGGER.finest(
          "sentenceCache = " + mapToString(this.resultCache.sentenceCache.asMap()),
        )
      }
    }
  }

  override fun activateDefaultFalseFriendRules() {
    val languageTool: JLanguageTool = (this.languageTool ?: return)

    // from JLanguageTool.activateDefaultFalseFriendRules (which is private)
    val falseFriendRulePath: String =
        JLanguageTool.getDataBroker().rulesDir + "/" + JLanguageTool.FALSE_FRIEND_FILE
    var exception: Exception? = null

    try {
      val falseFriendRules: List<AbstractPatternRule> =
          languageTool.loadFalseFriendRules(falseFriendRulePath)
      for (rule: Rule in falseFriendRules) languageTool.addRule(rule)
    } catch (e: IOException) {
      exception = e
    } catch (e: ParserConfigurationException) {
      exception = e
    } catch (e: SAXException) {
      exception = e
    }

    if (exception != null) {
      Logging.LOGGER.warning(
        I18n.format("couldNotLoadFalseFriendRules", exception, falseFriendRulePath),
      )
    }
  }

  override fun activateLanguageModelRules(languageModelRulesDirectory: String) {
    val languageTool: JLanguageTool = (this.languageTool ?: return)

    try {
      languageTool.activateLanguageModelRules(File(languageModelRulesDirectory))
    } catch (e: IOException) {
      Logging.LOGGER.warning(
        I18n.format("couldNotLoadLanguageModel", e, languageModelRulesDirectory),
      )
    }
  }

  override fun activateNeuralNetworkRules(neuralNetworkRulesDirectory: String) {
    val languageTool: JLanguageTool = (this.languageTool ?: return)

    try {
      languageTool.activateNeuralNetworkRules(File(neuralNetworkRulesDirectory))
    } catch (e: IOException) {
      Logging.LOGGER.warning(
        I18n.format("couldNotLoadNeuralNetworkModel", e, neuralNetworkRulesDirectory),
      )
    }
  }

  override fun activateWord2VecModelRules(word2vecRulesDirectory: String) {
    val languageTool: JLanguageTool = (this.languageTool ?: return)

    try {
      languageTool.activateWord2VecModelRules(File(word2vecRulesDirectory))
    } catch (e: IOException) {
      Logging.LOGGER.warning(I18n.format("couldNotLoadWord2VecModel", e, word2vecRulesDirectory))
    }
  }

  override fun enableRules(ruleIds: Set<String>) {
    val languageTool: JLanguageTool = (this.languageTool ?: return)

    // for strange reasons, there is no JLanguageTool.enableRules
    for (ruleId: String in ruleIds) {
      languageTool.enableRule(ruleId)
    }
  }

  override fun enableEasterEgg() {
    val languageTool: JLanguageTool = (this.languageTool ?: return)

    languageTool.addRule(
      object : Rule() {
        override fun getId(): String {
          return "bspline"
        }

        override fun getDescription(): String {
          return "Unknown basis function"
        }

        override fun match(sentence: AnalyzedSentence): Array<RuleMatch> {
          val matches = ArrayList<RuleMatch>()

          for (token: AnalyzedTokenReadings in sentence.tokens) {
            if (token.token.equals("hat", ignoreCase = true)) {
              matches.add(
                RuleMatch(
                  this,
                  sentence,
                  token.startPos,
                  token.endPos,
                  "Unknown basis function. Did you mean <suggestion>B-spline</suggestion>?",
                ),
              )
            }
          }

          return matches.toTypedArray()
        }
      },
    )

    languageTool.addRule(
      object : Rule() {
        override fun getId(): String {
          return "ungendered"
        }

        override fun getDescription(): String {
          return "Ungendered variant"
        }

        override fun match(sentence: AnalyzedSentence): Array<RuleMatch> {
          val matches = ArrayList<RuleMatch>()

          for (token: AnalyzedTokenReadings in sentence.tokens) {
            val tokenString: String = token.token

            if (
              (tokenString.length >= 2)
              && tokenString.substring(tokenString.length - 2).equals("er", ignoreCase = true)
            ) {
              matches.add(
                RuleMatch(
                  this,
                  sentence,
                  token.startPos,
                  token.endPos,
                  "Ungendered variant detected. "
                  + "Did you mean <suggestion>$tokenString*in</suggestion>?",
                ),
              )
            }
          }

          return matches.toTypedArray()
        }
      },
    )
  }

  companion object {
    private const val RESULT_CACHE_EXPIRE_AFTER_MINUTES = 60L

    fun mapToString(map: Map<*, *>?): String {
      if (map == null) return "null"
      val builder = StringBuilder("{")

      for ((key: Any?, value: Any?) in map) {
        if (builder.length > 1) builder.append(", ")
        appendObjectToBuilder(key, builder)
        builder.append(": ")
        appendObjectToBuilder(value, builder)
      }

      builder.append("}")
      return builder.toString()
    }

    private fun appendObjectToBuilder(obj: Any?, builder: StringBuilder) {
      if (obj != null) {
        builder.append("\"")
        builder.append(StringEscapeUtils.escapeJava(obj.toString()))
        builder.append("\"")
      } else {
        builder.append("null")
      }
    }
  }
}
